### ISSUE TITLE
Fix readline when input buffer is broken up into multiple calls

### ISSUE THREAD

**Issue by squeek502** [luvit/luvit#714]
From #709:

> There is still a lingering broken edge case when extremely large inputs are sent. If they are so large that the onKey call is broken up into multiple calls, then the cutoff point is arbitrary, meaning that it can end at something like '\027' and then the next call would start with '[D', which would then cause the trailing \027 to go unhandled and then the proceeding [D would be treated as normal characters and inserted into the Editor's line.

@creationix suggested the decoder style:
 * https://github.com/luvit/luvit/pull/709#issuecomment-97504664
 * https://github.com/luvit/luvit/pull/709#issuecomment-97505725

> Sounds good, but that's going to require a more robust control sequence parser to be able to reliably tell what is an unfinished control sequence waiting on more input and what isn't.

> For example, the F1 key sends '\027[[A' and is totally unhandled at the moment, so that would need to be treated as a delimeter rather than an unfinished control sequence. Otherwise, there'd be no way to differentiate it from something like a '\027[' that happens to be an unfinished '\027[D'.

Useful references:
* http://en.wikipedia.org/wiki/ANSI_escape_code#Sequence_elements
* http://en.wikipedia.org/wiki/C0_and_C1_control_codes

**Comment by squeek502** [comment]
Just realized that proper control sequence parsing would also be necessary for multiple-key inputs that have an unhandled control sequence within them.

For example, F1 + Left Arrow Key together would look like `\027[[A\027[D`, which would output: `Unhandled key(s): \027[[A\027[D` because no key handler consumes the `\027[[A` and onKey doesn't know how much of the string should be unhandled (it doesn't know when the F1 stops), so it just assumes the input is one unhandled key and treats it as such.

**Comment by creationix** [comment]
Do you have enough information to implement this yourself?  As much as I love writing decoders, I've got a lot of other stuff on my plate and should really behave and finish what I've already started.

**Comment by creationix** [comment]
Some example protocols in the decoder style can be found at:

 - https://github.com/luvit/lit/blob/master/deps/websocket-codec.lua - websocket
 - https://github.com/luvit/lit/blob/master/deps/http-codec.lua - http

**Comment by squeek502** [comment]
I'll take a crack at it, but might not end up starting until a day or two from now. The decoder style seems fairly straight forward; it's the control sequence parsing that might be a bit of an issue.

**Comment by squeek502** [comment]
Yeah, looks like this might not be very easy. The basic case is pretty simple: `\x1D[\x40-\x5F]` for single character escape sequences or (in regex) `CSI([\x20-\x2F]*)?((\d+;?)+)?([\x20-\x2F]*)?[\x40-\x7E]` where `CSI` is either `\x1D[` or `\xC2\x9B`.

However, there are a lot of weird cases. For example, in my terminal:
* F1: `\027[[A` (a non-standard/invalid control sequence, but [is a known sequence](http://en.wikipedia.org/wiki/ANSI_escape_code#Invalid_and_ambiguous_sequences_in_use))
* Shift + F1: `\027[23~` (no issue there)
* Alt + F1: `\027\027[[A` (Alt seems to prefix all control sequences with an extra ESC)
* Ctrl + F1: `\027[11^` (no issue there)

And as far as I can tell, other terminals handle that stuff differently.

So, this seems like it's going to get slightly messy.

More info:
* http://invisible-island.net/xterm/ctlseqs/ctlseqs.html
* http://vt100.net/docs/vt100-ug/chapter3.html#S3.3.3
* https://github.com/adoxa/ansicon/blob/1fe1d8f1305ef9ef7e8d8b57e3281c69a9e563e9/sequences.txt

Potentially useful reference implementations:
 * https://github.com/adoxa/ansicon/blob/master/ANSI.c#L1107-L1307 (seems to handle things like double escape sequences and private sequences like `\027[[A` properly)
 * https://github.com/cahirwpz/lua-carapace/blob/master/src/keydec.lua (might fail on things like double escape sequences and private sequences like `\027[[A`)

**Comment by creationix** [comment]
Another implementation we can look at is node's readline implementation https://github.com/joyent/node/blob/master/lib/readline.js

**Comment by squeek502** [comment]
Nice, node's implementation seems pretty complete but it makes heavy use of regex which might be hard/impossible to translate to Lua patterns.

**Comment by creationix** [comment]
I think if we just implement the deframer using the comment at https://github.com/joyent/node/blob/master/lib/readline.js#L936-L960 that will get us 90% of what we need right?

Sure a single regexp can't do that, but it shouldn't be too hard.

**Comment by squeek502** [comment]
Potentially, just note that `num ;` refers to an arbitrarily-sized semi-colon delimited list of numbers. That comment also seems to be missing `ESC [ ? num ; modifier letter` (not sure about the `; modifier` part of that) which [seems to exist as a private sequence](https://github.com/adoxa/ansicon/blob/master/sequences.txt#L119-L122).

I'll be writing up the test cases first and then I'll submit those as a WIP PR to provide a starting point.

**Comment by creationix** [comment]
Another useful decoder would be one that takes unicode chunks and looks for
partial characters at the end and buffers just that part will the rest of
the character comed down the pipe.

So suppose you have a message with a single byte character, 3 byte
character, and another single byte character:

```
ABBBC
````

And suppose it comes down the line as three chunks:

```
AB
B
BC
```

What you want is the multibyte character to never be broken and emit events
like:

```
A
BBBC
```

If we had such a decoder, it could be applied as well.  If the escape
sequences are all valid utf-8 characters, it could even be done before the
escape parser, but if not, it could be processed after the fact and only
process the plain-text bits.
